lib/src/backend.rs

```diff
@@ -473,6 +473,15 @@ pub trait Backend: Send + Sync + Debug {
 
     fn empty_tree_id(&self) -> &TreeId;
 
+    /// An estimate of how many concurrent requests this backend handles well. A
+    /// local backend like the Git backend (at until it supports partial clones)
+    /// may want to set this to 1. A cloud-backed backend may want to set it to
+    /// 100 or so.
+    ///
+    /// It is not guaranteed that at most this number of concurrent requests are
+    /// sent.
+    fn concurrency(&self) -> usize;
```

Comment:
This should return a `Option<usize>`, as it avoids sentinel values. As there probably are some backends which aren't concurrent or parallel. 

Sidenote: I had a something similar sketched out for `run`.

Reply:
I don't see "1" as a sentinel value here. It seems fair to say that for the git backend "how many concurrent requests this backend handles well" is 1. What do you think?

Reply:
> I don't see "1" as a sentinel value here. It seems fair to say that for the git backend "how many concurrent requests this backend handles well" is 1

I don't disagree with this statement but from an extensibility PoV, I'd like to have the option of saying that we can't be concurrent. Maybe someone has a stupid SVN backend or other internal remnants which just don't support this access pattern.

Reply:
> I'd like to have the option of saying that we can't be concurrent

Ah, we actually already do that in the Git backend by wrapping it in a Mutex. That's part of the reason a concurrency level above 1 doesn't make sense for it. Note that the `Backend` trait has async methods already, and we don't provide any promises that we won't make two concurrent calls to it even if this new function returns 1.

The on-disk representation of a commit store may also need some safety to protect against concurrent processes accessing it, but that should also be handled within the backend. Does that make sense?

Reply:
> Ah, we actually already do that in the Git backend by wrapping it in a Mutex. That's part of the reason a concurrency level above 1 doesn't make sense for it. Note that the `Backend` trait has async methods already, and we don't provide any promises that we won't make two concurrent calls to it even if this new function returns 1.
> 
> The on-disk representation of a commit store may also need some safety to protect against concurrent processes accessing it, but that should also be handled within the backend.

Ack, makes sense. I would rather consider this a "local" issue as we're just able to represent it better in the type system. I don't mind the concurrent requests in the backend itself, as everyone is free to synchronize in the backend impl.

Here are two versions of the issue I want to highlight. 

V1: 
```rust

// Sadly, still not stable. 
if backend.jobs.is_some() && let Some(jobs) = backend.jobs() {
   // Use our parallel impl. 
} 
// We can use the simple version, as we know jobs >= 1 canot happen. 
// ...
```

V2: 

```rust

let jobs = let Some(jobs) = backend.jobs() {
    jobs
} else {
 // Fallback to a single job, as we know the backend can't be parallel.  
    1 
};

if jobs == 1 {
   // Use the simple impl 
}

// Use the parallel impl 
```

Reply:
To clarify, the only reason we switch between two different algorithms is because I couldn't get the concurrent one to perform as well as the sequential one. If I had been able to make it as fast (or if we do in the future), we should delete the sequential version. So it's not that we *cannot* use the concurrent version with certain backends, it's just that it's (currently) a bit slower.